include/cell.tcc

```diff
@@ -949,3 +949,101 @@ Eigen::VectorXd mpm::Cell<Tdim>::interpolate_nodal_acceleration(
 
   return acceleration;
 }
+
+//! Assign velocity constraint
+//! Constrain directions can take values between 0 and Dim-1
+template <unsigned Tdim>
+bool mpm::Cell<Tdim>::assign_velocity_constraint(unsigned face_id, unsigned dir,
+                                                 double velocity) {
+  bool status = true;
+  try {
+    //! Constrain directions can take values between 0 and Dim * Nphases - 1
+    if (dir >= 0 && dir < Tdim) {
+      if (velocity_constraints_.find(face_id) == velocity_constraints_.end())
+        this->velocity_constraints_.insert(
+            std::make_pair<unsigned, std::vector<std::pair<unsigned, double>>>(
+                static_cast<unsigned>(face_id),
+                std::vector<std::pair<unsigned, double>>()));
+
+      this->velocity_constraints_[face_id].emplace_back(
+          std::make_pair<unsigned, double>(static_cast<unsigned>(dir),
+                                           static_cast<double>(velocity)));
+    } else
+      throw std::runtime_error("Constraint direction is out of bounds");
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+//! Apply velocity constraints
+template <unsigned Tdim>
+void mpm::Cell<Tdim>::apply_velocity_constraints() {
+
+  // TODO
+}
+
+//! Compute all face normal 3d
+template <>
+inline void mpm::Cell<2>::compute_normals() {
+
+  for (unsigned face_id = 0; face_id < 4; ++face_id) {
```

Comment:
could you `const` a variable to 4, which says why it is 4? Why can't it be 3? if it is a triangle?